In [1]:
#imports
#numpy,pandas,scipy, math, matplotlib
import numpy as np
import pandas as pd
import scipy
from math import sqrt
import matplotlib.pyplot as plt

In [2]:
#estimators
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn import linear_model

In [3]:
#model metrics
from sklearn.metrics import accuracy_score
from sklearn.metrics import r2_score
from sklearn.model_selection import cross_val_score


In [4]:
credit = pd.read_csv('creditChecked.csv', header =0)
credit.head()

Unnamed: 0  ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  SeptemberPay  \
0           0   1      20000    2          2         1   24             2   
1           1   2     120000    2          2         2   26            -1   
2           2   3      90000    2          2         2   34             0   
3           3   4      50000    2          2         1   37             0   
4           4   5      50000    1          2         1   57            -1   

   AugustPay  JulyPay  ...  JuneBill  MayBill  AprilBill  SeptemberPayment  \
0          2       -1  ...         0        0          0                 0   
1          2        0  ...      3272     3455       3261                 0   
2          0        0  ...     14331    14948      15549              1518   
3          0        0  ...     28314    28959      29547              2000   
4          0       -1  ...     20940    19146      19131              2000   

   AugustPayment  JulyPayment  JunePayment  MayPayment  AprilPayment  default  
0            689            0            0           0             0        1  
1           1000         1000         1000           0          2000        1  
2           1500         1000         1000        1000          5000        0  
3           2019         1200         1100        1069          1000        0  
4          36681        10000         9000         689           679        0  

[5 rows x 26 columns]

In [5]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 26 columns):
Unnamed: 0          30000 non-null int64
ID                  30000 non-null int64
LIMIT_BAL           30000 non-null int64
SEX                 30000 non-null int64
EDUCATION           30000 non-null int64
MARRIAGE            30000 non-null int64
AGE                 30000 non-null int64
SeptemberPay        30000 non-null int64
AugustPay           30000 non-null int64
JulyPay             30000 non-null int64
JunePay             30000 non-null int64
MayPay              30000 non-null int64
AprilPay            30000 non-null int64
SeptemberBill       30000 non-null int64
AugustBill          30000 non-null int64
JulyBill            30000 non-null int64
JuneBill            30000 non-null int64
MayBill             30000 non-null int64
AprilBill           30000 non-null int64
SeptemberPayment    30000 non-null int64
AugustPayment       30000 non-null int64
JulyPayment         30000 non-

All the features became int when I imported them from the EDA analysis, so I had to re-do the type change. 

In [6]:
credit = credit.astype({"ID":'category', "SEX":'category', "EDUCATION":'category', "AGE":'category', "MARRIAGE":'category', "SeptemberPay":'category', "AugustPay":'category', "JulyPay":'category', "JunePay":'category', "MayPay":'category', "AprilPay":'category', "default":'category'})

In [7]:
credit.drop('Unnamed: 0', axis = 1, inplace = True)

In [8]:
credit.drop(columns=['ID'])

LIMIT_BAL SEX EDUCATION MARRIAGE AGE SeptemberPay AugustPay JulyPay  \
0          20000   2         2        1  24            2         2      -1   
1         120000   2         2        2  26           -1         2       0   
2          90000   2         2        2  34            0         0       0   
3          50000   2         2        1  37            0         0       0   
4          50000   1         2        1  57           -1         0      -1   
...          ...  ..       ...      ...  ..          ...       ...     ...   
29995     220000   1         3        1  39            0         0       0   
29996     150000   1         3        2  43           -1        -1      -1   
29997      30000   1         2        2  37            4         3       2   
29998      80000   1         3        1  41            1        -1       0   
29999      50000   1         2        1  46            0         0       0   

      JunePay MayPay  ... JuneBill  MayBill  AprilBill  SeptemberPayment  \
0          -1     -2  ...        0        0          0                 0   
1           0      0  ...     3272     3455       3261                 0   
2           0      0  ...    14331    14948      15549              1518   
3           0      0  ...    28314    28959      29547              2000   
4           0      0  ...    20940    19146      19131              2000   
...       ...    ...  ...      ...      ...        ...               ...   
29995       0      0  ...    88004    31237      15980              8500   
29996      -1      0  ...     8979     5190          0              1837   
29997      -1      0  ...    20878    20582      19357                 0   
29998       0      0  ...    52774    11855      48944             85900   
29999       0      0  ...    36535    32428      15313              2078   

       AugustPayment  JulyPayment  JunePayment  MayPayment  AprilPayment  \
0                689            0            0           0             0   
1               1000         1000         1000           0          2000   
2               1500         1000         1000        1000          5000   
3               2019         1200         1100        1069          1000   
4              36681        10000         9000         689           679   
...              ...          ...          ...         ...           ...   
29995          20000         5003         3047        5000          1000   
29996           3526         8998          129           0             0   
29997              0        22000         4200        2000          3100   
29998           3409         1178         1926       52964          1804   
29999           1800         1430         1000        1000          1000   

       default  
0            1  
1            1  
2            0  
3            0  
4            0  
...        ...  
29995        0  
29996        0  
29997        1  
29998        1  
29999        1  

[30000 rows x 24 columns]

I dropped this Two columns, because they were of no use to the study

In [9]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
ID                  30000 non-null category
LIMIT_BAL           30000 non-null int64
SEX                 30000 non-null category
EDUCATION           30000 non-null category
MARRIAGE            30000 non-null category
AGE                 30000 non-null category
SeptemberPay        30000 non-null category
AugustPay           30000 non-null category
JulyPay             30000 non-null category
JunePay             30000 non-null category
MayPay              30000 non-null category
AprilPay            30000 non-null category
SeptemberBill       30000 non-null int64
AugustBill          30000 non-null int64
JulyBill            30000 non-null int64
JuneBill            30000 non-null int64
MayBill             30000 non-null int64
AprilBill           30000 non-null int64
SeptemberPayment    30000 non-null int64
AugustPayment       30000 non-null int64
JulyPayment         30000 non-null int

I selected the following feautures First: LIMIT_BAL, EDUCATION,  AGE 

In [14]:
#features
features = credit.iloc[:,[1,3,5]]
print('Summary of feature sample')
features.head()

Summary of feature sample


LIMIT_BAL EDUCATION AGE
0      20000         2  24
1     120000         2  26
2      90000         2  34
3      50000         2  37
4      50000         2  57

In [15]:
#dependent variable
depVar = credit['default']

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
X_train, X_test, y_train, y_test = train_test_split(features, depVar)

In [18]:
X_train.shape, X_test.shape

((22500, 3), (7500, 3))

In [19]:
#Models
modelSVM = SVC()
modelRF = RandomForestClassifier()
modelKNN = KNeighborsClassifier()

In [20]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

SVC Model Tuning

In [21]:
parameter_candidates = [
  {'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001], 'kernel': ['rbf']},
]
SVC_CV = GridSearchCV(modelSVM, parameter_candidates,scoring='accuracy', verbose=1, n_jobs=-1)
SVC_CV.fit(X_train,y_train)


C:\Users\albch\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


Fitting 3 folds for each of 8 candidates, totalling 24 fits


[Parallel(n_jobs=-1)]: Done  24 out of  24 | elapsed: 19.7min finished


GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='warn', n_jobs=-1,
             param_grid=[{'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001],
                          'kernel': ['rbf']}],
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring='accuracy', verbose=1)

In [22]:
print(SVC_CV.best_params_)
print(SVC_CV.best_score_)

{'C': 1, 'gamma': 0.001, 'kernel': 'rbf'}
0.7812


In [23]:
SVC_BestFit = SVC(C= 1, gamma= 0.001, kernel= 'rbf')

Random Forest Model Tuning

In [24]:
param_grid = { 
    'n_estimators': [200, 500],
    'max_features': ['auto'],
    'max_depth' : [4,5,6,7,8],
    'criterion' :['gini']
}
CV_rfc = GridSearchCV(estimator=modelRF, param_grid=param_grid, cv= 3, n_jobs=-1)
CV_rfc.fit(X_train, y_train)

#modelRF.fit(X_train,y_train)

GridSearchCV(cv=3, error_score='raise-deprecating',
             estimator=RandomForestClassifier(bootstrap=True, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators='warn', n_jobs=None,
                                              oob_score=False,
                                              random_state=None, verbose=0,
                                              warm_start=False),
             iid

In [25]:
print(CV_rfc.best_params_)
print(CV_rfc.best_score_)

{'criterion': 'gini', 'max_depth': 4, 'max_features': 'auto', 'n_estimators': 200}
0.7812444444444444


In [26]:
RF_BestFit = RandomForestClassifier(criterion= 'gini', max_depth= 4, max_features= 'auto', n_estimators= 200)

KNN Model Tuning

In [27]:
parameters = {'n_neighbors':[4,5,6,7],
              'leaf_size':[1,3,5],
              'algorithm':['auto', 'kd_tree'],
              'n_jobs':[-1]}

#Fit the model
KNNCV = GridSearchCV(modelKNN, param_grid=parameters,  verbose=1, n_jobs=-1)
KNNCV.fit(X_train,y_train)


C:\Users\albch\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


Fitting 3 folds for each of 24 candidates, totalling 72 fits


[Parallel(n_jobs=-1)]: Done  34 tasks      | elapsed:    2.5s
[Parallel(n_jobs=-1)]: Done  72 out of  72 | elapsed:    5.0s finished


GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30,
                                            metric='minkowski',
                                            metric_params=None, n_jobs=None,
                                            n_neighbors=5, p=2,
                                            weights='uniform'),
             iid='warn', n_jobs=-1,
             param_grid={'algorithm': ['auto', 'kd_tree'],
                         'leaf_size': [1, 3, 5], 'n_jobs': [-1],
                         'n_neighbors': [4, 5, 6, 7]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=1)

In [28]:
print(KNNCV.best_params_)
print(KNNCV.best_score_)

{'algorithm': 'auto', 'leaf_size': 5, 'n_jobs': -1, 'n_neighbors': 6}
0.7677777777777778


In [29]:
KNN_BestFit = KNeighborsClassifier(algorithm= 'auto', leaf_size= 5, n_jobs= -1, n_neighbors= 6)

Model Fitting

In [30]:
SVC_BestFit.fit(X_train, y_train)

SVC(C=1, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.001, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [31]:
RF_BestFit.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=4, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [32]:
KNN_BestFit.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=5, metric='minkowski',
                     metric_params=None, n_jobs=-1, n_neighbors=6, p=2,
                     weights='uniform')

Predictions

In [33]:
names = []
results = []

In [34]:
names.append('RF')
RF_predict = RF_BestFit.predict(X_test)

In [35]:
results.append(accuracy_score(y_test, RF_predict))
accuracy_score(y_test, RF_predict)

0.7714666666666666

In [36]:
names.append('SVM')
SVM_predict = SVC_BestFit.predict(X_test)

In [37]:
results.append(accuracy_score(y_test, SVM_predict))
accuracy_score(y_test, SVM_predict)

0.7714666666666666

In [38]:
names.append('KNN')
KNN_predict = KNN_BestFit.predict(X_test)

In [39]:
results.append(accuracy_score(y_test, KNN_predict))
accuracy_score(y_test, KNN_predict)

0.7589333333333333

In [40]:
from sklearn import metrics
print(metrics.classification_report(y_test, KNN_predict))

              precision    recall  f1-score   support

           0       0.78      0.97      0.86      5786
           1       0.33      0.05      0.09      1714

    accuracy                           0.76      7500
   macro avg       0.55      0.51      0.48      7500
weighted avg       0.67      0.76      0.68      7500



In [41]:

print(metrics.classification_report(y_test, SVM_predict))

              precision    recall  f1-score   support

           0       0.77      1.00      0.87      5786
           1       0.00      0.00      0.00      1714

    accuracy                           0.77      7500
   macro avg       0.39      0.50      0.44      7500
weighted avg       0.60      0.77      0.67      7500



C:\Users\albch\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [42]:

print(metrics.classification_report(y_test, RF_predict))

              precision    recall  f1-score   support

           0       0.77      1.00      0.87      5786
           1       0.00      0.00      0.00      1714

    accuracy                           0.77      7500
   macro avg       0.39      0.50      0.44      7500
weighted avg       0.60      0.77      0.67      7500



In [49]:
for x in range (0,3):
    print(str(names[x]) + ' accuracy value is: ' + str(results[x]) )
    

RF accuracy value is: 0.7714666666666666
SVM accuracy value is: 0.7714666666666666
KNN accuracy value is: 0.7589333333333333


Model behaviour very similar with an accuracy of 0.77 with this features:
LIMIT_BAL,  EDUCATION,  AGE

